lib/chef/provider/user/smartos.rb

```diff
@@ -0,0 +1,48 @@
+#
+# Author:: Dave Eddy (<dave@daveeddy.com>)
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the 'License');
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an 'AS IS' BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'chef/mixin/shell_out'
+require 'chef/provider/user/useradd'
+
+class Chef
+  class Provider
+    class User
+      class SmartOS < Chef::Provider::User::Useradd
+        provides :user, os: 'solaris2', platform_family: 'smartos'
+
+        def check_lock
+          shadow_line = shell_out!('getent', 'shadow', new_resource.username).stdout.strip
```

Comment:
Yeah, I don't think you'd get here actually. check_lock isn't called until the action is executed in the converge phase, but we should already have #define_resource_requirements from the user provider from which we inherit, which checks if the user exists and raises an error if it doesn't here: https://github.com/chef/chef/blob/master/lib/chef/provider/user.rb#L104

The exception isn't raised in whyrun though. Might be interesting to test what happens if the user doesn't exist and you run whyrun, it should say that the user would be locked based on the assumption.